better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1004 @@
+traditional. One has to weigh the risk of users reflexively writing
+`sort(areInOrder: <)`—against the value of these simplifications.  The
+argument label helps a bit, because *if* you happen to consider the
+equal elements case, you will see they clearly `areInOrder` but and
+realize `<` will return false.  But if you don't happen to consider
+that case, you won't get the sort you expected… sometimes.  Whether to
+break with precedent in order to get a simpler API with a clearer
+contract is an engineering decision you will have to make. To reduce
+the risk you could add this assertion, which will fire if the ordering
+is strict-weak:
+
+```
+assert(
+  self.isEmpty || areInOrder(first!, first!),
+  "Partial ordering required; did you pass a strict-weak ordering?")
+```
+
+
+### Project-Wide Documentation Policies
+
+The only reason you haven't seen
+complexity documented in this talk up to now is that I have a policy
+that operations have constant complexity unless specifically
+documented otherwise.
+
+Which brings me to this aside…  A big part of making the documentation
+problem tractable is having some well-chosen project-wide policies
+that save you from repeating common patterns. So I'm not going to
+prescribe your project's policy about this, but you should choose one,
+and write it down somewhere.  Your project's choice of policy can make
+the difference between documentation being useful and being burdensome
+or inconsistent (at which point people will just stop reading and
+writing it).
+
+For example,
+
+- Every declaration outside a function body must have a documentation comment that describes
+  its contract.
+  - Start with a summary sentence fragment.
+    - Describe what a function or method does and what it returns.
+    - Describe what a property or type is.
+    - Separate the fragment from any additional documentation with a blank line and end it
+      with a period.
+  - Preconditions, postconditions and invariants obviously implied by the summary need not
+    be explicitly documented.
+  - Declarations that fulfill protocol requirements are exempted when
+    nothing useful can be added to the documentation of the protocol
+    requirement itself.
+
+- End every file with a newline.
+- Do not strip trailing whitespace from lines you're not editing; it creates spurious VC diffs.
+- Document the performance of every operation that doesn't execute in constant time and space.
+
+
+It is reasonable to put information in the policies without which the
+project's other documentation would be incomplete or confusing, but
+you should be aware that it implies policies must be read.
+
+
+I want to mention one other thing: everything you see in these
+function signatures is implicitly part of the function's contract. For
+example, the signature if `sort` says the predicate must operate on
+arguments of type T, and return a `Bool`, so we didn't have to spell
+that out as a precondition in documentation.
```

Comment:
By consequence, the function's name is also implicitly part of the function's contract. I think this warrants further discussion, especially how it can facilitate omitting additional needless words.

Consider, for example, a function called "length()" on a container.

Reply:
If names can be incorrect, then they can be communicating information redundant with the contract documentation.

Consider Swift's Array.contains function:

```
func contains(Self.Element) -> Bool

Returns a Boolean value indicating whether the sequence contains the given element.
```

The contract is entirely implied by the name and signature of the function and, therefore, redundant.

Why are we okay with this redundancy when it occurs?

Reply:
> Why use two different wordings for the same thing

It is intentional.  When you use exactly the same wording, the documentation very often seems tautological, and saying a thing two different ways helps a reader be sure they understand what's being said.  Implementation, documentation, tests (and I guess name, too):  the idea is to say a thing different ways so you can validate one against the other.  Redundancy is necessary.